File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofServices/NFLPropBestBooksBundle.swift
import Foundation
import WagerproofModels

/// Backend-precomputed best-shop fields keyed by `player_id|market`.
/// Shipped in the app bundle until `nfl_dryrun_props` carries these columns.
struct NFLPropBestBooksRecord: Decodable {
    let bestOverBook: String?
    let bestOverBookName: String?
    let bestOverBookLogo: String?
    let bestOverLine: Double?
    let bestOverPrice: Double?
    let bestUnderBook: String?
    let bestUnderBookName: String?
    let bestUnderBookLogo: String?
    let bestUnderLine: Double?
    let bestUnderPrice: Double?

    enum CodingKeys: String, CodingKey {
        case bestOverBook = "best_over_book"
        case bestOverBookName = "best_over_book_name"
        case bestOverBookLogo = "best_over_book_logo"
        case bestOverLine = "best_over_line"
        case bestOverPrice = "best_over_price"
        case bestUnderBook = "best_under_book"
        case bestUnderBookName = "best_under_book_name"
        case bestUnderBookLogo = "best_under_book_logo"
        case bestUnderLine = "best_under_line"
        case bestUnderPrice = "best_under_price"
    }
}

enum NFLPropBestBooksBundle {
    static let index: [String: NFLPropBestBooksRecord] = load()

    private static func load() -> [String: NFLPropBestBooksRecord] {
        guard let url = Bundle.module.url(
            forResource: "nfl_dryrun_prop_best_books",
            withExtension: "json"
        ) else { return [:] }
        return (try? Data(contentsOf: url))
            .flatMap { try? JSONDecoder().decode([String: NFLPropBestBooksRecord].self, from: $0) }
            ?? [:]
    }
}

File: wagerproof-android-native/core/services/src/main/kotlin/com/wagerproof/core/services/NFLPropBestBooksBundle.kt
package com.wagerproof.core.services

import com.wagerproof.core.models.serialization.WagerproofJson
import com.wagerproof.core.shared.AppGroup
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Backend-precomputed best-shop fields keyed by `player_id|market`.
 * Shipped in the app assets until `nfl_dryrun_props` carries these columns.
 */
@Serializable
internal data class NFLPropBestBooksRecord(
    @SerialName("best_over_book") val bestOverBook: String? = null,
    @SerialName("best_over_book_name") val bestOverBookName: String? = null,
    @SerialName("best_over_book_logo") val bestOverBookLogo: String? = null,
    @SerialName("best_over_line") val bestOverLine: Double? = null,
    @SerialName("best_over_price") val bestOverPrice: Double? = null,
    @SerialName("best_under_book") val bestUnderBook: String? = null,
    @SerialName("best_under_book_name") val bestUnderBookName: String? = null,
    @SerialName("best_under_book_logo") val bestUnderBookLogo: String? = null,
    @SerialName("best_under_line") val bestUnderLine: Double? = null,
    @SerialName("best_under_price") val bestUnderPrice: Double? = null,
)

internal object NFLPropBestBooksBundle {

    /** Lazy so the assets read happens off the class-load path; failure → empty index. */
    val index: Map<String, NFLPropBestBooksRecord> by lazy { load() }

    private fun load(): Map<String, NFLPropBestBooksRecord> = runCatching {
        AppGroup.context.assets
            .open("nfl_dryrun_prop_best_books.json")
            .use { it.readBytes().decodeToString() }
            .let { WagerproofJson.decodeFromString<Map<String, NFLPropBestBooksRecord>>(it) }
    }.getOrDefault(emptyMap())
}
